pkg/registry/core/service/allocator/bitmap_test.go

```diff
@@ -23,116 +23,402 @@ import (
+					}
+				}
+				calls := sets.NewInt()
+				m.ForEach(func(i int) {
+					calls.Insert(i)
+				})
+				if len(calls) != len(ts) {
+					t.Errorf("[%d] expected %d calls, got %d", i, len(ts), len(calls))
+				}
+				if !calls.Equal(ts) {
+					t.Errorf("[%d] expected calls to equal testcase: %v vs %v", i, calls.List(), ts.List())
+				}
+			}
+		})
+	}
+}
+
+func TestSnapshotAndRestore(t *testing.T) {
+	testCases := []struct {
+		name      string
+		allocator func(max int, rangeSpec string) *AllocationBitmap
+		max       int
+	}{
+		{
+			name:      "NewAllocationMap",
+			allocator: NewAllocationMap,
+			max:       32,
+		},
+		{
+			name:      "NewAllocationReservedMap max < 64",
+			allocator: NewAllocationReservedMap,
+			max:       32},
+		{
+			name:      "NewAllocationReservedMap max > 64",
+			allocator: NewAllocationReservedMap,
+			max:       128,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			m := tc.allocator(tc.max, "test")
+			offset := 3
 			if ok, _ := m.Allocate(offset); !ok {
-				t.Errorf("[%d] error allocate offset %v", i, offset)
+				t.Errorf("error allocate offset %v", offset)
 			}
-			if !m.Has(offset) {
-				t.Errorf("[%d] expect offset %v allocated", i, offset)
+			spec, bytes := m.Snapshot()
+
+			m2 := tc.allocator(10, "test")
+			err := m2.Restore(spec, bytes)
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+
+			if m2.count != 1 {
+				t.Errorf("expect count to %d, but got %d", 0, m.count)
+			}
+			if !m2.Has(offset) {
+				t.Errorf("expect offset %v allocated", offset)
 			}
-		}
-		calls := sets.NewInt()
-		m.ForEach(func(i int) {
-			calls.Insert(i)
 		})
-		if len(calls) != len(tc) {
-			t.Errorf("[%d] expected %d calls, got %d", i, len(tc), len(calls))
+	}
+
+}
+
+// TestAllocateMaxReserved should allocate first values greater or equal than defaultReservedValues=32
+func TestAllocateMax_BitmapReserved(t *testing.T) {
+	max := 128
+	// just to double check off by one errors
+	allocated := 0
+	// modify if necessary
+	m := NewAllocationReservedMap(max, "test")
+	for i := 0; i < max-defaultReservedValues; i++ {
+		if _, ok, _ := m.AllocateNext(); !ok {
+			t.Fatalf("unexpected error")
 		}
-		if !calls.Equal(tc) {
-			t.Errorf("[%d] expected calls to equal testcase: %v vs %v", i, calls.List(), tc.List())
+		allocated++
+	}
+
+	if f := m.Free(); f != defaultReservedValues {
+		t.Errorf("expect to get %d, but got %d", defaultReservedValues-1, f)
+	}
+
+	for i := 0; i < defaultReservedValues; i++ {
+		if m.Has(i) {
+			t.Errorf("unexpected allocated value %d", i)
 		}
 	}
+	// it should allocate the first value of the reserved block
+	if _, ok, _ := m.AllocateNext(); !ok {
+		t.Fatalf("unexpected error")
+	}
+	allocated++
+	if !m.Has(0) {
+		t.Errorf("expected allocated value 0")
+	}
+	if allocated != m.count {
+		t.Errorf("expect to get %d, but got %d", allocated, m.count)
+	}
+
+	if m.count != max-defaultReservedValues+1 {
```

Comment:
this is why I've added the allocated++ var, to be sure `max-defaultReservedValues+1` is correct